src/mysql_conn.erl

```diff
@@ -553,10 +553,22 @@ handle_info(ping, #state{socket = Socket, sockmod = SockMod} = State) ->
     #ok{} = mysql_protocol:ping(SockMod, Socket),
     setopts(SockMod, Socket, [{active, once}]),
     {noreply, schedule_ping(State)};
-handle_info({tcp_closed, _Socket}, State) ->
+handle_info({tcp_closed, Socket}, #state{socket = Socket} = State) ->
```

Comment:
Hm, I'm afraid I'm not a huge fan of adding special artificial code for tests. It always carries a risk to behave differently than the real thing. This spot here is pretty obvious, but as a rule of thumb I try to avoid it.

That said, I just realized that I can't do this bogus thing anyway, since it will then call `setopts` with the `fake_socket`. I'll think of something else :)

Reply:
What if we use a fake socket module? I think there are some tests that do that actually. Yes, look at `test/mock_tcp.erl`.

Reply:
You could also extract the socket from the connection state with `sys:get_state/1` and `element(4, State)` (3rd position in the state record, plus the record name tag), then pass it in the bogus message like `tcp, Socket, <<"Foo\n">>}`. Then, check that `active` is set to `once`, alternatively send another bogus message like `{tcp_closed, Socket}` and confirm that the connection process exits, or both.

Reply:
Use `sys:get_state/1` in the test? That's a good idea. No pollution in the actual implementation.

Using `element(4, State)` in the test is also fine by me. If the record ever changes the position of the socket element, the test would catch that.

Reply:
Nice, done :)